plugin/remote/stash/stash.go

```diff
@@ -0,0 +1,296 @@
+		AccessTokenURL:        r.URL + "/plugins/servlet/oauth/access-token",
+		CallbackURL:           httputil.GetScheme(req) + "://" + httputil.GetHost(req) + "/api/auth/stash.atlassian.com",
+		ConsumerKey:           r.Secret,
+		ConsumerPrivateKeyPem: r.PrivateKey,
+	}
+
+	// get the oauth verifier
+	verifier := req.FormValue("oauth_verifier")
+	if len(verifier) == 0 {
+		// Generate a Request Token
+		requestToken, err := consumer.RequestToken()
+		if err != nil {
+			return nil, err
+		}
+
+		// add the request token as a signed cookie
+		httputil.SetCookie(w, req, "stash_token", requestToken.Encode())
+
+		url, _ := consumer.AuthorizeRedirect(requestToken)
+		http.Redirect(w, req, url, http.StatusSeeOther)
+		return nil, nil
+	}
+
+	// remove stash token data once before redirecting
+	// back to the application.
+	defer httputil.DelCookie(w, req, "stash_token")
+
+	// get the tokens from the request
+	requestTokenStr := httputil.GetCookie(req, "stash_token")
+	requestToken, err := oauth1.ParseRequestTokenStr(requestTokenStr)
+	if err != nil {
+		return nil, err
+	}
+
+	// exchange for an access token
+	accessToken, err := consumer.AuthorizeToken(requestToken, verifier)
+	if err != nil {
+		return nil, err
+	}
+
+	// create the Stash client
+	var client = stash.New(
+		r.URL,
+		r.Secret,
+		accessToken.Token(),
+		accessToken.Secret(),
+		r.PrivateKey,
+	)
+
+	// get the currently authenticated Stash User
+	user, err := client.Users.Current()
+	if err != nil {
+		return nil, err
+	}
+
+	// put the user data in the common format
+	login := model.Login{
+		Login:  user.Username,
+		Access: accessToken.Token(),
+		Secret: accessToken.Secret(),
+		//Name:   user.DisplayName,
+	}
+
+	return &login, nil
+}
+
+// GetKind returns the internal identifier of this remote Stash instance.
+func (r *Stash) GetKind() string {
+	return model.RemoteStash
+}
+
+// GetHost returns the url.Host of this remote system.
+func (r *Stash) GetHost() string {
+	uri, _ := url.Parse(r.URL)
+	return uri.Host
+}
+
+// GetRepos fetches all repositories that the specified
+// user has access to in the remote system.
+func (r *Stash) GetRepos(user *model.User) ([]*model.Repo, error) {
+	var repos []*model.Repo
+	var client = stash.New(
+		r.URL,
+		r.Secret,
+		user.Access,
+		user.Secret,
+		r.PrivateKey,
+	)
+
+	// parse the hostname from the stash url
+	var stashurl, err = url.Parse(r.URL)
+	if err != nil {
+		return nil, err
+	}
+
+	// parse the hostname from the stash api
+	stashapi, err := url.Parse(r.API)
+	if err != nil {
+		return nil, err
+	}
+
+	list, err := client.Repos.List()
+	if err != nil {
+		return nil, err
+	}
+
+	var remote = r.GetKind()
+	var hostname = r.GetHost()
+
+	for _, item := range list {
+		// for now we only support git repos
+		if item.ScmId != "git" {
+			continue
+		}
+
+		// these are the urls required to clone the repository
+		var clone = fmt.Sprintf("https://%s%s/%s/%s.git", stashurl.Host, stashurl.Path, item.Project.Key, item.Name)
```

Comment:
I've tested this and did not work for me!

You are forcing https but it can be http also... you already have the stash url. just use the protocol from there.

Also `/scm/` is missing from the url. I think stash adds always that?